### DISCUSSION TITLE
fully differentiable ray tracing

### DISCUSSION THREAD

**Discussion by fighting-xq** [jeertmans/DiffeRT2d#89] [Q&A] [answered]
I just curious why continuous validity check leads to a fully differentiable ray-tracing?  The validity array just serves as an index, why can its smoothing lead to fully differentiable? 
`if is_true(valid, approx=approx):
      yield (tx_key, rx_key, path, path_candidate)`

**Comment by jeertmans** [comment]
Hello! I guess the line you are referring to is from `Scene.all_valid_paths`?

https://github.com/jeertmans/DiffeRT2d/blob/5fc52f01631f25b670c10430b24366cc10231cdf/differt2d/scene.py#L1211-L1215

The `all_valid_paths` function is a more convenient wrapper for plotting purpose, as path validity is represented by a [`Truthy`](https://differt2d.eertmans.be/latest/reference/differt2d.logic.html#differt2d.logic.Truthy) value, which can be any real number between 0 and 1.

The [`is_true`](https://differt2d.eertmans.be/latest/reference/differt2d.logic.html#differt2d.logic.is_true) check actually breaks the "fully differentiable ray tracing", as it creates a discontinuity in the output. However, if you look at the [power optimization example](https://differt2d.eertmans.be/latest/examples_gallery/plot_power_optimize.html), it uses `accumulate_over_paths` (see code below), which **does not** check for validity, but rather multiply the output value by the path validity (which is continuously differentiable).

https://github.com/jeertmans/DiffeRT2d/blob/5fc52f01631f25b670c10430b24366cc10231cdf/differt2d/scene.py#L1257-L1266

So, in the end, we do not perform any *check* during optimization, but only for plotting purposes. Did I answer your question?

**Reply by fighting-xq** [reply]
Thank you very much. It helps me a lot!!!

**Reply by fighting-xq** [reply]
I have another question: why does this approach give correct gradients? You know, you weight all the path with a validity matrix. But, there are many originally invalid paths being calculated, even with a small weight. So I just curious why this fully differentiable approach gives correct gradients.

**Reply by jeertmans** [reply, marked as answer]
What do you mean by "correct"?

* If you mean "numerically", then they are *correct* assuming JAX does compute the correct gradient, but this was already the case *without* smoothing;
* If you mean *physically*, then they are most likely incorrect, because the smoothing technique is a numerical trick introduced to allow continuously differentiable Ray Tracing. I will very likely investigate any possible physical link (e.g., with UTD coefficients) in a future paper, but this is still ongoing work.

In the end, the smoothing technique is mainly applied to make optimization easier, but it is disabled  when computing the exact RT solution.

**Reply by fighting-xq** [reply]
Thanks a lot!!! You are doing a great job!

**Reply by jeertmans** [reply]
You're welcome, and thanks!
